code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,233 @@
+	var/image/YouAreHere
+
+/obj/machinery/holomap/Initialize(mapload)
+	. = ..()
+	if(!display_map)
+		display_map = SSmapping.map_datum.technical_name
+	var/icon/base_icon = make_map()
+	var/icon/alpha_mask = icon('icons/effects/480x480.dmi', "scanline") //Scanline effect.
+	base_icon.AddAlphaMask(alpha_mask)
+	holomap_projection = image(base_icon, loc = src, layer = ABOVE_HUD_PLANE, pixel_x = -240, pixel_y = -240)
+	holomap_projection.plane = ABOVE_HUD_PLANE
+	holomap_projection.filters = filter(type = "drop_shadow", size = 8, color = "#ffffff65")
+	YouAreHere = image('icons/effects/holomap_icons.dmi', src, "YouAreHere", 22.1 , 2, -256+(x*2), -256+(y*2))
+	YouAreHere.plane = 22.1
+
+/obj/machinery/holomap/attack_hand(mob/user)
+	. = ..()
+	if(stat & (NOPOWER|BROKEN) || user.incapacitated() || !user.Adjacent(src))
+		return
+
+	if(!ishuman(user))
+		atom_say("Unable to locate subject's optical organ. Place optical organ infront of device and try again.")
+		return
+
+	if(emagged)
+		var/mob/living/carbon/M = user
+		M.flash_eyes(2, 1)
+	if(locate(user) in display_targets)
+		stopdisplaymap(user)
+	else
+		displaymap(user)
+
+/obj/machinery/holomap/proc/displaymap(mob/M)
+	if(!M.client)
+		return
+	display_targets.Add(M)
+	M.client.images += YouAreHere
+	M.client.images += holomap_projection
+	update_icon()
+	if(do_after(M, 5 MINUTES, FALSE, src, FALSE))
+		atom_say("Maximum display time reached, shutting down session.")
+		stopdisplaymap(M)
+	else
+		stopdisplaymap(M)
+
+/obj/machinery/holomap/proc/stopdisplaymap(mob/user)
+	display_targets.Remove(user)
+	user.client.images -= holomap_projection
+	user.client.images -= YouAreHere
+	update_icon()
+
+/obj/machinery/holomap/update_icon()
+	. = ..()
+
+	if(stat & (NOPOWER|BROKEN))
+		icon_state = "map_off"
+		return
+
+	if(length(display_targets))
+		icon_state = "map_projecting"
+	else if(emagged)
+		icon_state = "map_emag"
+	else
+		icon_state = "map"
+
+/obj/machinery/holomap/emag_act(user as mob)
+	emagged = TRUE
+	to_chat(user, "You short out the emission regulators of [src]!")
+	for(var/mob/living/carbon/M in display_targets)
+		M.flash_eyes(2, 1)
+
+/obj/machinery/holomap/power_change()
+	. = ..()
+	update_icon()
+
+/obj/machinery/holomap/proc/make_map()
+	var/icon/bigassicon = icon('icons/effects/480x480.dmi',"blank")
+	for(var/x_index in 1 to 240)
+		for(var/y_index in 1 to 240)
+			var/turf/T = locate(x_index, y_index, 2)
+			var/area/A = get_area(T)
+			var/icon_to_use = 'icons/effects/holomap_parts.dmi'
+			var/icon_state_to_use
+			var/dir_to_use = 2
+			var/list/directions_to_aliens = list()
+
+			if(isfloorturf(T) && !locate(/obj/structure/window) in T)
+				icon_state_to_use = "floor"
+
+			else if(iswallturf(T) || istype(T, /turf/simulated/mineral) || locate(/obj/structure/window) in T)
+				var/list/cardinals = list(NORTH, SOUTH, EAST, WEST)
+				var/number_of_solid_turfs_neighbors = 0
+				var/space_found = FALSE
+				icon_state_to_use = "full"
+				for(var/i in 1 to 4)
```

Comment:
This should be able to be simplified to this, with a few additional changes
```suggestion
				for(var/i in cardinals)
```

Reply:
Even better, you could use
```suggestion
				for(var/direction in GLOB.cardinal)
```
instead. (Don't forget to change the other places that are called `i` to `direction`